homeassistant/components/intent/__init__.py

```diff
@@ -218,46 +208,52 @@ async def async_handle(self, intent_obj: intent.Intent) -> intent.IntentResponse
         if "state" in slots:
             state_names = set(slots["state"]["value"])
 
-        states = list(
-            intent.async_match_states(
-                hass,
-                name=entity_name,
-                area=area,
-                domains=domains,
-                device_classes=device_classes,
-                assistant=intent_obj.assistant,
-            )
-        )
-
-        _LOGGER.debug(
-            "Found %s state(s) that matched: name=%s, area=%s, domains=%s, device_classes=%s, assistant=%s",
-            len(states),
-            entity_name,
-            area,
-            domains,
-            device_classes,
-            intent_obj.assistant,
+        match_constraints = intent.MatchTargetsConstraints(
+            name=entity_name,
+            area_name=area_id,
+            floor_name=floor_id,
+            domains=domains,
+            device_classes=device_classes,
+            assistant=intent_obj.assistant,
         )
-
-        if entity_name and (len(states) > 1):
-            # Multiple entities matched for the same name
-            raise intent.DuplicateNamesMatchedError(
-                name=entity_text or entity_name,
-                area=area_name or area_id,
+        match_result = intent.async_match_targets(hass, match_constraints)
+        if (not match_result.is_match) and (
+            match_result.no_match_reason
+            in (
+                intent.MatchFailedReason.DUPLICATE_NAME,
+                intent.MatchFailedReason.INVALID_AREA,
+                intent.MatchFailedReason.INVALID_FLOOR,
+            )
+        ):
+            # Don't try to answer questions for certain errors.
+            # Other match failure reasons are OK.
+            raise intent.MatchFailedError(
+                result=match_result, constraints=match_constraints
             )
 
         # Create response
         response = intent_obj.create_response()
         response.response_type = intent.IntentResponseType.QUERY_ANSWER
```

Comment:
We can get here if there if `match_result.is_match` is False. Is that expected?

Reply:
You now filter out expected errors. How can we avoid not checking this code when adding a new match failed reason?

Reply:
It is expected to get here with a failed match. If you ask "how many lights are on in the kitchen" and there are no lights on, you will technically have a failed match.

There are two kinds of reasons here: ones for entities failing to match, and others for bad area names, etc. I've moved the code to a method on the `MatchFailedReason` enum so it lives in a central place.